var/spack/repos/builtin/packages/py-cffi/package.py

```diff
@@ -22,29 +22,36 @@ class PyCffi(PythonPackage):
     version("1.15.0", sha256="920f0d66a896c2d99f0adbb391f990a84091179542c205fa53ce5787aff87954")
     version("1.14.6", sha256="c9a875ce9d7fe32887784274dd533c57909b7b1dcadcc128a2ac21331a9765dd")
     version("1.14.3", sha256="f92f789e4f9241cd262ad7a555ca2c648a98178a953af117ef7fad46aa1d5591")
-    version("1.13.0", sha256="8fe230f612c18af1df6f348d02d682fe2c28ca0a6c3856c99599cdacae7cf226")
-    version("1.12.2", sha256="e113878a446c6228669144ae8a56e268c91b7f1fafae927adc4879d9849e0ea7")
-    version("1.11.5", sha256="e90f17980e6ab0f3c2f3730e56d1fe9bcba1891eeea58966e89d352492cc74f4")
-    version("1.10.0", sha256="b3b02911eb1f6ada203b0763ba924234629b51586f72a21faacc638269f4ced5")
-    version("1.1.2", sha256="390970b602708c91ddc73953bb6929e56291c18a4d80f360afa00fad8b6f3339")
+    with default_args(deprecated=True):
+        version(
+            "1.13.0", sha256="8fe230f612c18af1df6f348d02d682fe2c28ca0a6c3856c99599cdacae7cf226"
+        )
+        version(
+            "1.12.2", sha256="e113878a446c6228669144ae8a56e268c91b7f1fafae927adc4879d9849e0ea7"
+        )
+        version(
+            "1.11.5", sha256="e90f17980e6ab0f3c2f3730e56d1fe9bcba1891eeea58966e89d352492cc74f4"
+        )
+        version(
+            "1.10.0", sha256="b3b02911eb1f6ada203b0763ba924234629b51586f72a21faacc638269f4ced5"
+        )
+        version("1.1.2", sha256="390970b602708c91ddc73953bb6929e56291c18a4d80f360afa00fad8b6f3339")
 
-    depends_on("c", type="build")  # generated
+    depends_on("c", type="build")
 
-    # ./spack-src/cffi/ffiplatform.py has _hack_at_distutils which imports
-    # setuptools before distutils, but only on Windows. This could be made
-    # unconditional to support Python 3.12
-    depends_on("python@:3.11", type=("build", "run"))
-
-    # python 3.12 support was released in @1.16:, however the removal
-    # in python3.12 of distutils has resulted in an imperfect fix for prefix-based
-    # tools like spack, see:
-    # https://github.com/spack/spack/pull/46224
-    # https://github.com/cython/cython/pull/5754#issuecomment-1752102480
-    # until this is correctly fixed, do not enable 3.12 support
-    # depends_on("python@:3.12", type=("build", "run"), when="@1.16:")
+    # Based on PyPI wheel availability
+    with default_args(type=("build", "link", "run")):
+        depends_on("python@3.8:3.13", when="@1.17:")
+        depends_on("python@3.8:3.12", when="@1.16")
+        depends_on("python@:3.11", when="@1.15.1")
+        depends_on("python@:3.10", when="@1.15.0")
+        depends_on("python@:3.9", when="@1.14.3:1.14")
+        depends_on("python@:3.8", when="@1.13:1.14.2")
+        depends_on("python@:3.7", when="@:1.12")
```

Comment:
again request to split forward and backward compat rules.
```suggestion
        depends_on("python@3.8:", when="@1.16:")

        depends_on("python@:3.13")
        depends_on("python@:3.12", when="@:1.16")
        depends_on("python@:3.11", when="@:1.15")
        depends_on("python@:3.10", when="@:1.15.0")
        depends_on("python@:3.9", when="@:1.14")
        depends_on("python@:3.8", when="@:1.14.2")
        depends_on("python@:3.7", when="@:1.12")
```

Reply:
Isn't this a bit conservative though? I think we only need the known `@:3.11` upperbound due to distutils breakage. Otherwise nothing?

Reply:
It's definitely quite conservative, but for low-level libraries that directly bind to CPython internals, there are quite a few breakages between minor versions. For example, https://github.com/python-cffi/cffi/pull/24 fixes support for Python 3.13. There are likely similar changes for 3.11, 3.10, etc. If someone wants to loosen these bounds later (e.g., if Python 3.14 is released and there is no 3.14 compatible release yet) then that's fine, otherwise I think it's safer to remain conservative.

Reply:
fine by me